piq/haarpsi.py

```diff
@@ -0,0 +1,204 @@
+"""PyTorch implementation of Haar Wavelet-Based Perceptual Similarity Index (HaarPSI)
+
+Reference:
+    [1] R. Reisenhofer, S. Bosse, G. Kutyniok & T. Wiegand (2017)
+        A Haar Wavelet-Based Perceptual Similarity Index for Image Quality Assessment
+        http://www.math.uni-bremen.de/cda/HaarPSI/publications/HaarPSI_preprint_v4.pdf
+    [2] Code from authors on MATLAB and Python
+        https://github.com/rgcda/haarpsi
+"""
+
+import functools
+from typing import Optional, Union
+
+import torch
+import torch.nn.functional as F
+from torch.nn.modules.loss import _Loss
+
+from piq.utils import _adjust_dimensions, _validate_input
+from piq.functional import similarity_map, rgb2yiq, haar_filter
+
+
+def haarpsi(x: torch.Tensor, y: torch.Tensor, reduction: Optional[str] = 'mean',
+            data_range: Union[int, float] = 1., scales: int = 3, subsample: bool = True,
+            c: float = 30.0, alpha: float = 4.2) -> torch.Tensor:
+    r"""Compute Haar Wavelet-Based Perceptual Similarity
+    Input can by greyscale tensor of colour image with RGB channels order.
+    Args:
+        x: Tensor of shape :math:`(N, C, H, W)` holding an distorted image.
+        y: Tensor of shape :math:`(N, C, H, W)` holding an target image
+        reduction: Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the sum of the output will be divided by the number of
+            elements in the output, ``'sum'``: the output will be summed.
+        data_range: The difference between the maximum and minimum of the pixel value,
+            i.e., if for image x it holds min(x) = 0 and max(x) = 1, then data_range = 1.
+            The pixel value interval of both input and output should remain the same.
+        scales: Number of Haar wavelets used for image decomposition.
+        subsample: Flag to apply average pooling before HaarPSI computation. See [1] for details.
+        c: Constant from the paper. See [1] for details
+        alpha: Exponent used for similarity maps weightning. See [1] for details
+
+    Returns:
+        HaarPSI : Wavelet-Based Perceptual Similarity between two tensors
+    
+    References:
+        [1] R. Reisenhofer, S. Bosse, G. Kutyniok & T. Wiegand (2017)
+            'A Haar Wavelet-Based Perceptual Similarity Index for Image Quality Assessment'
+            http://www.math.uni-bremen.de/cda/HaarPSI/publications/HaarPSI_preprint_v4.pdf
+        [2] Code from authors on MATLAB and Python
+            https://github.com/rgcda/haarpsi
+    """
+
+    _validate_input(input_tensors=(x, y), allow_5d=False, scale_weights=None)
+    x, y = _adjust_dimensions(input_tensors=(x, y))
+
```

Comment:
Add assertion to validate the size of the image to be greater than max possible `kernel size = 2 ** (scale + 1)`